crates/rethnet_evm/src/runtime.rs

```diff
@@ -0,0 +1,118 @@
+use std::{fmt::Debug, sync::Arc};
+
+use rethnet_eth::B256;
+use revm::{BlockEnv, CfgEnv, ExecutionResult, SpecId, TxEnv};
+
+use crate::{
+    blockchain::{AsyncBlockchain, SyncBlockchain},
+    db::{AsyncDatabase, SyncDatabase},
+    evm::build_evm,
+    inspector::RethnetInspector,
+    trace::Trace,
+    State,
+};
+
+/// The asynchronous Rethnet runtime.
+pub struct Rethnet<E>
+where
+    E: Debug + Send + 'static,
+{
+    blockchain: Arc<AsyncBlockchain<Box<dyn SyncBlockchain<E>>, E>>,
+    db: Arc<AsyncDatabase<Box<dyn SyncDatabase<E>>, E>>,
+    cfg: CfgEnv,
+}
+
+impl<E> Rethnet<E>
+where
+    E: Debug + Send + 'static,
+{
+    /// Constructs a new [`Rethnet`] instance.
+    pub fn new(
+        blockchain: Arc<AsyncBlockchain<Box<dyn SyncBlockchain<E>>, E>>,
+        db: Arc<AsyncDatabase<Box<dyn SyncDatabase<E>>, E>>,
+        cfg: CfgEnv,
+    ) -> Self {
+        Self {
+            blockchain,
+            db,
+            cfg,
+        }
+    }
+
+    /// Runs a transaction without committing the state.
+    pub async fn dry_run(
+        &mut self,
+        transaction: TxEnv,
+        mut block: BlockEnv,
+    ) -> (ExecutionResult, State, Trace) {
+        let blockchain = self.blockchain.clone();
+        let db = self.db.clone();
+        let cfg = self.cfg.clone();
+
+        if cfg.spec_id > SpecId::MERGE && block.prevrandao.is_none() {
```

Comment:
I had to add this as we're not setting the `mixHash` for block headers in Hardhat. This feels like a hack though. What do you think, @fvictorio ?

Reply:
I'm not sure I understand. We do set the `mixHash` in block headers.

With respect to this code... I think we should panic if we are in a post-merge hardfork and `prevrandao` is not set? I'm not 100% sure though, and I'm fine with leaving it as it is.

Reply:
There are about 300 tests (potential half due to the existence of two providers) that are currently failing without this.

If the requirement isn't met I'd just return an `Err`. Panicking would cause a crash so should only happen when an internal assumption about the data model isn't met. E.g. like the `state_root.unwrap()` we discussed the other day.

Reply:
https://github.com/NomicFoundation/hardhat/pull/3443